### PR TITLE
[FIX: #357] Change shuffle algorithm to Fisher-Yates shuffle

### PR DESCRIPTION
**Pull request by billmei**
## Description

**Resolves** #357

See the issue #357 for more information.

**Lodash#shuffle** #100 -> https://lodash.com/docs/4.17.4#shuffle 

## What does your PR belong to?
- [ ] Website
- [x] Snippets
- [ ] General / Things regarding the repository (like CI Integration)

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (non-breaking improvement of a snippet)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## [Lodash Backlog](https://github.com/Chalarangelo/30-seconds-of-code/issues/100)

- [ ] I added the prefix [UPDATE: `method.md`] or [ADD: `method.md`]
- [ ] I have referenced the `method` to the lodash backlog.

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have checked that the changes are working properly
- [ ] I have checked that there isn't any PR doing the same
- [ ] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: snippets/shuffle.md [hunk 1]
```diff
@@ -1,10 +1,20 @@
 ### shuffle
 
-Randomizes the order of the values of an array.
+Randomizes the order of the values of an array, in place.
 
-Use `Array.sort()` to reorder elements, using `Math.random()` in the comparator.
+Users the Fisher-Yates reorder the elements, based on the [Lodash implementation](https://github.com/lodash/lodash/blob/b2ea6b1cd251796dcb5f9700c4911a7b6223920b/shuffle.js).
```
Comment (skatcat31) [inline snippets/shuffle.md]: Users?
Is it supposed to be `Uses the Fisher-Yates algoritm to reorder the elements of the array, based on the [Lodash implimentation](https://github.com/lodash/lodash/blob/b2ea6b1cd251796dcb5f9700c4911a7b6223920b/shuffle.js)`

**Comment by Chalarangelo** [comment]
I will review and merge this later when I am on my laptop, as I am currently on mobile. This is a very welcome change, thanks for contributing!

**Comment by atomiks** [comment]
Here's a shorter version (I think0

```js
const shuffle = ([...arr]) => {
  let m = arr.length;
  while (m) {
    const i = Math.floor(Math.random() * m--);
    [arr[m], arr[i]] = [arr[i], arr[m]];
  }
  return arr;
};
```

**Comment by atomiks** [comment]
For fun, to check for bias:

```js
const determineBias = shuffle => {
  const frequencies = {
    '123': 0,
    '132': 0,
    '213': 0,
    '231': 0,
    '321': 0,
    '312': 0
  }

  for (let i = 0; i < 100000; i++) {
    frequencies[shuffle([1,2,3]).reduce((acc, n) => acc + n, '')]++
  }

  return frequencies
}
```

```js
determineBias(([...arr]) => {
  let m = arr.length;
  while (m) {
    const i = Math.floor(Math.random() * m--);
    [arr[m], arr[i]] = [arr[i], arr[m]];
  }
  return arr;
})
// {123: 16536, 132: 16663, 213: 16975, 231: 16540, 312: 16834, 321: 16452}

determineBias(arr => [...arr].sort(() => .5 - Math.random()))
// {123: 24947, 132: 12503, 213: 25149, 231: 12627, 312: 12363, 321: 12411}
```

**Comment by billmei** [comment]
Thanks everyone! 💯

**Comment by skatcat31** [comment]
@Pl4gue why the `[...arr]`? isn't that destructing it into an array... from an array?

**Comment by Pl4gue** [comment]
@skatcat31 No, actually not. The spread operator in this case spreads the given array into an array to be edited in the function. You can either spread the array before passing it to the function or spread it in the parameter like in this snippet. I have to admit that it's some kind of double-destructing :joy: 
Do you understand what I mean?

**Comment by skatcat31** [comment]
ah, you're sanitizing it as a "pure" function. Got it.

**Comment by billmei** [comment]
I noticed that since this was converted from an in-place shuffle into a pure function, the documentation is no longer accurate. I created a new PR to update this: https://github.com/Chalarangelo/30-seconds-of-code/pull/363

**Comment by atomiks** [comment]
@skatcat31 it's "rest" destructuring, to create a clone in the param.

```js
const [...arr] = [1, 2, 3, 4];
// `arr` is an array with the values of the array
```
So when given the func is given an argument it does that.

**Pull request merged**